File: sliding window/java/MaxFruitCountOf2Types.java
import java.util.HashMap;

public class MaxFruitCountOf2Types {

    public static void main(String[] args) {
        
        System.out.println(findLength(new char [] {'A', 'B', 'C', 'A', 'C'}));
        System.out.println(findLength(new char [] {'A', 'B', 'C', 'B', 'B', 'C'}));

    }

    public static int findLength(char[] arr) {
        
        HashMap<Character, Integer> map = new HashMap<>();

        int windowStart = 0;
        int result = Integer.MIN_VALUE;
        int windowLength = 0;

        for(int windowEnd = 0; windowEnd < arr.length; windowEnd++){
            char c = arr[windowEnd];
            map.put(c, map.getOrDefault(c, 0) + 1);
            windowLength++;

            while(map.size() > 2){
                char k = arr[windowStart];
                map.put(k, map.getOrDefault(k, 0) - 1);

                if(map.get(k) <= 0)
                    map.remove(k);

                windowStart++;    
                windowLength--;
            }

            result = Math.max(result, windowLength);
        
        }


        return result;
    }
    
}

File: sliding window/typescript/MaxFruitCountOf2Types.ts
function totalFruit(fruits: number[]): number {

    let windowStart = 0;
    let result = 0;
    let fruitLengthMap = new Map<number, number>();

    for (let windowEnd = 0; windowEnd < fruits.length; windowEnd++) {
        fruitLengthMap.set(fruits[windowEnd], (fruitLengthMap.get(fruits[windowEnd]) ?? 0) + 1);

        while (fruitLengthMap.size > 2) {
            let fruit = fruitLengthMap.get(fruits[windowStart]) ?? 0;
            fruit > 1 ?
                fruitLengthMap.set(fruits[windowStart], (fruitLengthMap.get(fruits[windowStart]) ?? 0) - 1)
                :
                fruitLengthMap.delete(fruits[windowStart]);

            windowStart++;
        }

        result = Math.max(result, windowEnd - windowStart + 1);
    }

    return result;

};

console.log(totalFruit([1, 2, 3, 2, 2]));
console.log(totalFruit([0, 1, 2, 2]));
console.log(totalFruit([3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4]));